src/zenml/integrations/azure/secrets_managers/azure_secrets_manager.py

```diff
@@ -0,0 +1,205 @@
+from typing import Any, ClassVar, Dict, List
+
+from azure.identity import DefaultAzureCredential
+from azure.keyvault.secrets import SecretClient
+
+from zenml.exceptions import SecretExistsError
+from zenml.logger import get_logger
+from zenml.secret.base_secret import BaseSecretSchema
+from zenml.secret.secret_schema_class_registry import SecretSchemaClassRegistry
+from zenml.secrets_managers.base_secrets_manager import BaseSecretsManager
+
+logger = get_logger(__name__)
+
+ZENML_SCHEMA_NAME = "zenml-schema-name"
+ZENML_GROUP_KEY = "zenml-group-key"
+
+
+def prepend_group_name_to_keys(secret: BaseSecretSchema) -> Dict[str, str]:
+    """This function adds the secret group name to the keys of each
+    secret key-value pair to allow using the same key across multiple
+    secrets.
+    Args:
+        secret: The ZenML Secret schema
+    """
+    return {f"{secret.name}-{k}": v for k, v in secret.content.items()}
+
+
+def remove_group_name_from_key(combined_key_name: str, group_name: str) -> str:
+    """This function serves to remove the secret group name from the secret
+    key.
+    Args:
+        combined_key_name: Full name as it is within the Azure secrets manager
+        group_name: Group name (the ZenML Secret name)
+    Returns:
+        The cleaned key
+    """
+    if combined_key_name.startswith(f"{group_name}-"):
+        return combined_key_name[len(f"{group_name}-") :]
+    else:
+        raise RuntimeError(
+            f"Key-name `{combined_key_name}` does not have the "
+            f"prefix `{group_name}`. Key could not be "
+            f"extracted."
+        )
+
+
+class AzureSecretsManager(BaseSecretsManager):
+    """Class to interact with the Azure secrets manager.
+    Attributes:
+        project_id:  This is necessary to access the correct Azure project.
+                     The project_id of your Azure project space that contains
+                     the Secret Manager.
+    """
+
+    key_vault_name: str
+
+    # Class configuration
+    FLAVOR: ClassVar[str] = "azure_secrets_manager"
+    CLIENT: ClassVar[Any] = None
+
+    @classmethod
+    def _ensure_client_connected(cls, vault_name: str) -> None:
+        if cls.CLIENT is None:
+            KVUri = f"https://{vault_name}.vault.azure.net"
+
+            credential = DefaultAzureCredential()
+            cls.CLIENT = SecretClient(vault_url=KVUri, credential=credential)
+
+    def register_secret(self, secret: BaseSecretSchema) -> None:
```

Comment:
I wasn't sure where to write this, but we noticed yesterday that azure converts '_' to '-' - we should either raise an exception here or log a warning else they will save a 'secret_name' and then get confused when get('secret_name') returns nothing.

Reply:
I could just add a check for that, like this:

```python
if "_" in secret.name:
            raise ValueError(
                f"The secret name `{secret.name}` contains an underscore. "
                f"This will cause issues with Azure. Please try again."
            )
```